### ISSUE TITLE
Controlled TextField allows modification apart from value prop

### ISSUE THREAD

**Issue by jdhuntington** [OfficeDev/office-ui-fabric-react#1350]
#### Priorities and help requested (not applicable if asking question):

Are you willing to submit a PR to fix? *Yes*

This is an important use-case for me, but I'm happy to build my own control and copy the styling as a workaround.

#### Describe the issue:

A _controlled_ `TextField` can be modified by a user to contain contents not matching the value.

#### Actual behavior:

In this code, user interaction with the `TextField` causes the text on screen to be modified and become out of sync with the data in the `value` prop.

    <Fabric.TextField label='Fabric - Controlled by value' value='static value' />

#### Expected behavior:

In this code, I expect user interaction with the `TextField` to have no effect.

    <Fabric.TextField label='Fabric - Controlled by value' value='static value' />

For contrast, this JSX which renders a vanilla DOM `input` node does not allow user modification.

    <input id='vanilla-textfield' value='static value' type='text' />

### If applicable, please provide a codepen repro:

https://codepen.io/anon/pen/XMBevW

---

Would a PR changing this behavior be welcome?

Additionally, the delay between user modification and the invocation of the `onChanged` callback poses several issues - would a change to that default behavior be welcome as well?

**Comment by dzearing** [comment]
@jdhuntington Sounds good to me! I agree with your expectations :)

**Comment by jordandrako** [comment]
@jdhuntington @dzearing Do you still want this behavior, as it would be a breaking change? If so, I'm thinking the default input should not allow user modification with the `value` attribute, and instead use `defaultValue` to enable TextField's `onChange` method?

**Comment by jdhuntington** [comment]
I'd still find it valuable, though I've worked around it for now and unfortunately haven't taken the time to implement this.

**Comment by jordandrako** [comment]
@jdhuntington My implementation definitely wasn't good, and I haven't had time to look at this since. Do you feel you could submit a PR making the workaround you've found a more official change?

**Comment by JasonGore** [comment]
I've confirmed this is [still an issue](https://codepen.io/jasongore/pen/NOOaQb) and, as mentioned earlier, a resolution to this issue will also be perceived as a breaking change by some. For these reasons I've created #6827 as a resolution and will mark this issue blocked on that one.

**Comment by ecraig12345** [comment]
I talked with @dzearing about this on Friday and he wants me to go ahead and make the change, which I'll start working on shortly. So I guess speak now if you have major objections? :)

Meanwhile, here's yet another repro pen I made while thinking through the issue. 
https://codepen.io/ecraig12345/pen/xQgyMW

**Comment by ecraig12345** [comment]
After working through this, I think the "right" change is to make the TextField follow much stricter controlled vs uncontrolled behavior similar to React's `<input>`--see the description of my PR #7187 for what all that includes. Trouble is that would *definitely* be a breaking change for the (likely many) people who are using the component incorrectly (I found quite a few examples of this within Fabric). 

@JasonGore suggested putting the new behavior behind a flag for now and making it the default in v7, so the latest PR iteration does that. The flag is currently called `strictMode` but I'm open to naming suggestions. I may follow a similar pattern with other components that have this type of issue, for example #5326 for SpinButton.

@jdhuntington FYI in case you're still interested in this.

**Comment by ecraig12345** [comment]
If anyone finds this issue and wants a fix for it before 7.0, here's something you can do (assumes your field's value is stored in `this.state.value`):
```
private _onChange: (ev: any, newValue?: string) => {
  if (/*new value is valid*/) {
    this.setState({ value: newValue });
  } else {
    this.setState({ value: this.state.value });
  }
}
```

**Comment by starkWebDev** [comment]
The workaround provided on Office UI Fabrics website, as well as the ones suggested here do not work in MobX. Despite setting the value to its current value the user is still able to continue typing. Though it does not change the state value, it does allow them to type.
Any tips on a workaround when using MobX stores for state?

**Comment by ecraig12345** [comment]
Very sorry about the delay here! I finally have a fix for this in #8723, and if all goes well, that should be included in Fabric 7.

**Comment by ecraig12345** [comment]
Just checked in the fix in Fabric 7!
